From aa885b11d4cf9347da706f5842ed5b83842b51d9 Mon Sep 17 00:00:00 2001
From: tchalivendra <tchalivendra@maxlinear.com>
Date: Thu, 21 Nov 2024 12:27:52 +0530
Subject: [PATCH] WLANRTSYS-85223 Hostapd is crashing during reboot/factory
 reset as interfaces weren't initialized when SIGHUP is triggered from PWHM.

Null pointer check has been added to handle this scenario in hostapd.
Fixup to WLANRTSYS-47651: MXL proprietary NL80211_BAND_ATTR_HT_TX_BF_CAPA

Signed-off-by: Tejaswi <tchalivendra@maxlinear.com>
---
 src/ap/hw_features.c       | 3 +++
 src/ap/mxl_drv_callbacks.c | 4 ++++
 2 files changed, 7 insertions(+)

diff --git a/src/ap/hw_features.c b/src/ap/hw_features.c
index 4a95a8be3..0c186771a 100644
--- a/src/ap/hw_features.c
+++ b/src/ap/hw_features.c
@@ -790,6 +790,9 @@ static int ieee80211n_supported_ht_tx_bf_capab(struct hostapd_iface *iface)
 #ifdef CONFIG_VENDOR_MXL
 int mxl_ieee80211n_supported_ht_tx_bf_capab(struct hostapd_iface *iface)
 {
+	if (!iface->current_mode || is_6ghz_freq(iface->freq) || !iface->conf->ieee80211n) {
+		return 0;
+	}
 	return ieee80211n_supported_ht_tx_bf_capab(iface);
 }
 #endif /* CONFIG_VENDOR_MXL */
diff --git a/src/ap/mxl_drv_callbacks.c b/src/ap/mxl_drv_callbacks.c
index 6f6756ed1..021f4934d 100644
--- a/src/ap/mxl_drv_callbacks.c
+++ b/src/ap/mxl_drv_callbacks.c
@@ -98,11 +98,15 @@ static void mxl_hostapd_event_ml_csa_done_on_reporting_link(struct hostapd_data
 		return;
 	}
 
+	max_chan_switch_time = sibling_hapd->iconf->mxl_conf.max_chan_switch_time;
 	if (event == EVENT_CHAN_SWITCH_COUNT_DONE) {
+		ret = mxl_ml_send_bss_critical_update_info(hapd,
+				ML_BCN_PER_STA_PROF_MAX_SWITCH_TIME, 0, max_chan_switch_time);
 		if (ret) {
 			wpa_printf(MSG_ERROR, "CSA Max Channel Switch Time update on reporting link failed");
 			return;
 		}
+		hapd->mxl_data.ml_ie_max_chan_sw_time_ie_len_add = true;
 	}
 
 	ieee802_11_set_beacon_per_bss_only(hapd);
-- 
2.43.0

